### ISSUE TITLE
[US] Dockerize PostGOAT

### ISSUE THREAD

**Issue by Alexicazana** [WillTrem/UInnovate#182]
| MoSCoW      |  Risk               | Story Points  |
| ------------- |------------- | ------------- |
| Must |   High                   |        5           |

## Risks
- Requires reorganization of part of the stack's architecture (microservice-adjacent ish)
- Need to have solid understanding of Docker - creating an image to publish on Dockerhub and to be used by others (fault-free)
- Can be hard to define the expected behaviour since we are not yet super clear on the other moving parts (eg Postgoat website documentation and deployment, containerizing the application so that an application configurator may use it appropriately; how does an app configurator truly expect to use the platform?)

## Pre-requisites
- [x] Meet with the stakeholder to clarify the requirements
- [x] Understand the types of users that will be involved - define their roles, what permissions do they have?
- [x] Read Dockerization documentation (see links below) to ensure proper creation and deployment (risk of having to fix this later can have heavy consequences)

## Tasks
- [x] Read Docker docs about dockerizing a full stack app (https://milanwittpohl.com/projects/tutorials/Full-Stack-Web-App/dockerizing-our-front-and-backend) 
- [x] Create Dockerfile to define dependencies, base image, and any necessary configurations
- [x] Create a self-contained docker image that has the front end deployed on it as well as all the required tooling (.sh scripts, and volume access to the "usecases") (created from dockerfile that uses an official base image + everything else needed)

**Comment by Alexicazana** [comment]
Technically, the new dockerfile *should* be correctly written, but there is a lot of confusion relating to the way the new dockerfile should overlap with the existing dockerfile for the other three services using containers. The dockerfile as is should technically live in the root of the application platform, but there's a bit of confusion with the compose.yml file... The other issue is that we have to copy the services existing on the development side for the application configurators, which raises a few concerns for me/creates a bit of messiness.

Note that I had a couple of issues relating to the other containers, which need to be replicated for the application configurator.

TLDR; I'm still figuring out how I should package this new docker service so that I can publish it to Dockerhub and then just churn out use cases by using this custom image. In the meantime, here's the dockerfile as is, so that there's at least some sense of progress visible:

![Screenshot 2024-01-12 at 9 23 37 PM](https://github.com/WillTrem/UInnovate/assets/55753055/60707f6e-fe92-414f-83db-3b18ef74fbc0)

**Comment by Alexicazana** [comment]
Second compose.yml with the new postgoat service added:

![Screenshot 2024-01-12 at 9 24 25 PM](https://github.com/WillTrem/UInnovate/assets/55753055/486f92a0-4534-4e66-9725-90a846432f49)

This needs additional integration and testing, as well as a better understanding on how to implement alongside existing compose.yml (existing containers).

**Comment by Alexicazana** [comment]
Last minute changes made to existing containers, need to resolve and account for this in new compose.yml for the new service

![Screenshot 2024-01-12 at 9 27 46 PM](https://github.com/WillTrem/UInnovate/assets/55753055/0148bb6e-05a5-4c17-81e8-1abc803c725e)

**Comment by Alexicazana** [comment]
Updates:
- Had to change dockerfile into a multi stage build; ran into a bunch of issues when trying to run npm run build (tsconfig and viteconfig were missing some stuff and had a couple of errors in them like the build target version; this took an entire day to debug as you can read in these GitHub issues about:
https://github.com/vitejs/vite/issues/6985
https://github.com/remix-run/remix/issues/7969
). This along with having to modify parts of files that simply weren't respecting TypeScript rules and for which the dev environment was manually forced to allow, but this was not working during the build process obviously.
-I'm getting somewhere but keep running into the following issue:
![Screenshot 2024-01-23 at 11 19 25 PM](https://github.com/WillTrem/UInnovate/assets/55753055/45063c0c-d1a7-4102-a3fd-57dcf8e256a6)

Working on a couple of fixes; here's a few breadcrumbs relating to the issues I am having:
https://github.com/moby/buildkit/issues/4305
https://github.com/docker/buildx/issues/1079

^Most recent update (4 days ago) on the issue...

**Comment by Alexicazana** [comment]
On the other hand, after almost two days of debugging, I finally managed to run a build to create JS out of all of our TS/TSX files locally! :D
![Screenshot 2024-01-23 at 11 26 56 PM](https://github.com/WillTrem/UInnovate/assets/55753055/53a7412a-53e0-4a38-83b0-ff65417af18a)

This was the first step, because the image I have to create needs to be able to do this in a container, so it was necessary to get this working locally first.

**Comment by Alexicazana** [comment]
MAJOR UPDATES!
- The build files get created successfully, allowing the static web page to eventually be deployed using the minified js files inside of a folder called `/build`

![Screenshot 2024-01-26 at 12 25 19 AM](https://github.com/WillTrem/UInnovate/assets/55753055/03d16f33-632b-47c1-b2a8-839c600a6ad5)

- I finally managed to successfully create the image container:
![Screenshot 2024-01-25 at 11 49 32 PM](https://github.com/WillTrem/UInnovate/assets/55753055/f0bb07d6-f628-46c7-9dd7-a77921284894)

- Got the container running!
![Screenshot 2024-01-26 at 12 19 02 AM](https://github.com/WillTrem/UInnovate/assets/55753055/5f7586d8-6ef2-4afc-abd6-cd9c8b707597)

Only issue is that it seems that the content of the web page is not displaying when I open localhost:8080 (it does however open correctly):
![Screenshot 2024-01-26 at 12 18 55 AM](https://github.com/WillTrem/UInnovate/assets/55753055/82b6f50a-0c98-4408-bce9-2a34ee56469b)

**Comment by lyubageorgieva** [comment]
This US is pushed to Iteration 9, since we are missing the demo video, as our Stakeholder requested to keep doing the videos for sign-off.

**Comment by TheOrangeDuke** [comment]
Looks good! Signing off!

**Comment by Alexicazana** [comment]
Included original software architecture diagram I had done for this story:

<img width="1106" alt="081e2bcc-4efa-45d4-8a75-50fb2bdcbf14" src="https://github.com/WillTrem/UInnovate/assets/55753055/c6e38c08-6cdc-453b-95a7-53a1a94e37b4">
